lib/internal/fs/promises.js

```diff
@@ -663,19 +665,46 @@ async function writeFile(path, data, options) {
   options = getOptions(options, { encoding: 'utf8', mode: 0o666, flag: 'w' });
   const flag = options.flag || 'w';
 
-  if (!isArrayBufferView(data)) {
-    validateStringAfterArrayBufferView(data, 'data');
-    data = Buffer.from(data, options.encoding || 'utf8');
+  if (isStream(data) || isIterable(data)) {
```

Comment:
Streams are async iterable, should we simplify this?
```suggestion
  if (isIterable(data)) {
```